# spikequant
Remove and Quantify Organism Spike Reads from Sequencing Data

## Prerequisites

Must have snakemake available in `PATH`.

### dependencies
Either execute snakemake with `--use-singularity` (or the updated syntax for snakemake 8), a snakemake profile with singularity configured, or from a conda environment with the following dependencies precent:

- pigz
- ART read simulator
- spades
- metawrap
- seqkit


## Test Data
Test data can be generated by running the following:

```
snakemake --snakefile testdata/Snakefile --directory $PWD/testdata/data/
```

configure the types of data it creates via the config file in testdata/config.yaml.  We have commited a 100k read dataset to the repo, along with its assembly and binning results. Due to the poor quality of this low-depth assembly, we don't bother running the bin refinement step of metawrap.

## Run `spikequant` on the testdata

```
snakemake --directory testresults --snakefile workflow/Snakefile --config spiketable=$PWD/benchmarking/manifests/raw.csv
```

## Benchmarking
The benchmarking workflow compares alignment methods, off-target database, and spike preprocessing methods to identify optimal usage parameters.  Details of the benchmarking can be found in [benchmarking/README.md](benchmarking/README.md)].
